transport/tchannel/outbound_test.go

```diff
@@ -47,6 +48,108 @@ var newOutbounds = []func(*tchannel.Channel, string) transport.Outbound{
 	},
 }
 
+func TestOutboundHeaders(t *testing.T) {
+	tests := []struct {
+		context context.Context
+		headers transport.Headers
+
+		wantHeaders []byte
+		wantError   string
+	}{
+		{
+			context: yarpc.WithBaggage(context.Background(), "FOO", "Bar"),
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x0B, 'c', 'o', 'n', 't', 'e', 'x', 't', '-', 'f', 'o', 'o', // context-foo
+				0x00, 0x03, 'B', 'a', 'r',
+			},
+		},
+		{
+			headers: transport.Headers{"contextfoo": "bar"},
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x0A, 'c', 'o', 'n', 't', 'e', 'x', 't', 'f', 'o', 'o',
+				0x00, 0x03, 'b', 'a', 'r',
+			},
+		},
+		{
+			headers: transport.Headers{"Foo": "bar"},
+			wantHeaders: []byte{
+				0x00, 0x01,
+				0x00, 0x03, 'f', 'o', 'o',
+				0x00, 0x03, 'b', 'a', 'r',
+			},
+		},
+		{
+			headers:   transport.Headers{"context-foo": "bar"},
+			wantError: `application headers cannot start with "Context-"`,
+		},
+	}
+
+	for _, tt := range tests {
+		server := testutils.NewServer(t, nil)
+		defer server.Close()
+		hostport := server.PeerInfo().HostPort
+
+		server.GetSubChannel("service").SetHandler(tchannel.HandlerFunc(
+			func(ctx context.Context, call *tchannel.InboundCall) {
+				var headers []byte
+
+				err := tchannel.NewArgReader(call.Arg2Reader()).Read(&headers)
+				if assert.NoError(t, err, "failed to read request headers") {
+					assert.Equal(t, tt.wantHeaders, headers, "headers did not match")
+				}
+
+				var body []byte
+				err = tchannel.NewArgReader(call.Arg3Reader()).Read(&body)
+				if assert.NoError(t, err, "failed to read request body") {
+					assert.Equal(t, []byte("world"), body)
+				}
```

Comment:
nit: suggest adding a test utility in yarpc that looks like the terribly named `ReadArgsV2`:
https://godoc.org/github.com/uber/tchannel-go/raw#ReadArgsV2

That will make code like this a little cleaner, since you can do:
```go
headers, body, err := readArgs(call)
if assert.NoError(t, err, "failed to read args") {
  // verify body and headers
}
```

the test utility method can wrap errors with info about whether headers or body failed to read as well. (tchannel's version does not do this since we need to keep the same error type)